plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/ColumnIdentity.java

```diff
@@ -48,10 +50,14 @@ public ColumnIdentity(
         this.id = id;
         this.name = requireNonNull(name, "name is null");
         this.typeCategory = requireNonNull(typeCategory, "typeCategory is null");
-        this.children = requireNonNull(children, "children is null");
+        requireNonNull(children, "children is null");
         checkArgument(
                 children.isEmpty() == (typeCategory == PRIMITIVE),
                 "Children should be empty if and only if column type is primitive");
+        this.children = new LinkedHashMap<>(children.size());
+        for (ColumnIdentity child : children) {
+            this.children.put(child.getId(), child);
```

Comment:
- validate no duplicates (`.put` doesn't override) and the put value isn't null (`rnn` on `child`)
- you should wrap the final map in the `unmodifiableMap`  to make it effectively immutable
- `ImmutableMap` also maintains insertion order and `ImmutableMap.builder` does all the above for you automatically


- relying on map order is quite error-prone; can we do better?

Reply:
@findepi I think ultimately it'd the best if it's possible to use https://github.com/trinodb/trino/pull/8754, and get rid of this structure altogether. But I haven't tried the two PRs together to see if `Schema` method helps with simplifying this at all places. 

It looks like we need to the order when we translate a dereference into a subcolumn (and back).

One option is to keep Type structure as single source of truth at all places, but that'd require us to translate between (`RowType.Fields` ordering --> name --> ColumnIdentity's Id) and back. that seems a bit clunky.

Another option is to encode ordering in ColumnIdentity, but use something like 

```
    private final Map<Integer, ColumnIdentity> childrenById;
    private final List<Integer> idsForPositions;
```

and provide methods like `getChildAtPosition` and `getChildById`.  `getChildren()` method will return unordered map.

Reply:
@findepi do you have any reservations against the second option above? I think it'd be atleast less error-prone than using `ImmutableMap` or `linkedinhashmap`. (but slightly clunkier during creation)

Reply:
> Another option is to encode ordering in ColumnIdentity, but use something like
> 
> ```
>     private final Map<Integer, ColumnIdentity> childrenById;
>     private final List<Integer> idsForPositions;
> ```

works for me.

or perhaps simpler:
```java
private final List<ColumnIdentity> children;
private final Map<Integer, ColumnIdentity> childrenById;
```

> I think it'd be atleast less error-prone than using `ImmutableMap` or `linkedinhashmap`.

I realized we do not return the map in any getter, so the current code (using ImmutableMap) works for me too.